Enhancement: incremental backup with CBT

ShellyKa13 opened the pull request:
**What this PR does / why we need it**:
This VEP proposes to add incremental backup with CBT using Libvirt and QEMU capabilities.
The current backup options rely on CSI-storage and create a full snapshot of the storage with each backup. This leads to longer backup times and increased storage usage. Additionally, when stored off-site, a large amount of data needs to be copied over which further extends the backup time. This can be improved by leveraging QEMU's capability to create incremental backups, which use Changed Block Tracking (CBT) to save only the changes made since the last backup.


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # https://github.com/kubevirt/enhancements/issues/25

**Special notes for your reviewer**:

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Proposal for incremental backup with CBT
```

aglitke commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+
+According to the previous solution, if the VM was offline when the backup was requested, we will start the VM in a paused state. When the VM is in a paused state, there is no issue with starting it during the backup process, as Libvirt and QEMU guarantee that any VM operations, such as overwriting blocks in the data file, will not affect the backup's consistency. This consistency assurance could  ...
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup.
```
I guess we could pause the VM and then stop if once the VMBackup resource is deleted.

aglitke replied:
We should address here what happens if a VM crashes.  In this case, the CBT information may be corrupted and must be discarded.  Qemu offers an API to check if the dirty bitmaps are clean.  We should check this at VM startup when repopulating libvirt with the checkpoint information.  If a corrupted bitmap is found then we will discard the checkpoints and the next backup will need to be a full backup.

ShellyKa13 replied:
Yeah I mentioned it as one of the implementation phases but sure Ill add a paragraph describing this

aglitke commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+    <disk type='file' device='disk' model='virtio-non-transitional'>
+      <driver name='qemu' type='qcow2' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/run/kubevirt-private/libvirt/qemu/swtpm/qcow2overlay.qcow2' index='2'>
+        <dataStore type='file'>
```
Will it be type='block' in case of a block mode PVC?

ShellyKa13 replied:
yes. Ill mention that this is an example and that the type should be as of the type of the base raw image

aglitke commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,330 @@
+```xml
+    <disk type='file' device='disk' model='virtio-non-transitional'>
+      <driver name='qemu' type='qcow2' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/run/kubevirt-private/libvirt/qemu/swtpm/qcow2overlay.qcow2' index='2'>
````
Better to use the volume name here because a VM could have multiple disks and therefore multiple overlays.

aglitke commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+Optional. Should be either `push` or `pull`. Initially, as mentioned, only `push` will be allowed. If not specified, `push` will be the default behavior.
+- `incremental`<br>
+Optional. If not specified, a full backup will be taken. When set, the field value should be the checkpoint name from which the incremental backup should be taken.
+
```
Recommending to add a pvcName field here.  For push, it will write the backup data to this PVC (instead of creating a PVC.  For pull mode, this PVC would be used for the scratch space.

How will kubevirt decide to create the PVC if it must?  Which storage class would be used for example?

ShellyKa13 replied:
maybe the same as defined for the vm state?

ShellyKa13 replied:
Or do you think we should use the storage spec in cdi or something like that?

aglitke replied:
I think it's complicated to get right and it's much easier to leave this task to the backup client.  In the future we could add automatic PVC provisioning.  Choosing a correct size for the PVC will also be harder since it can depend on many factors that only the backup client is privy to (ie. time since last backup, policy, etc).

alicefr replied:
@ShellyKa13 can you please define who is creating this cr? Backup vendor?Guest owner?

aglitke commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+> Note: All of the information mentioned below are suggestions for backup vendors or users to fetch the backup. **There will not be an API provided for this.**
+
+**Push mode:**
+PVC name containing the backup is provided in the VirtualMachineBackup CR status. A data-mover pod can be spawned and copy the data over to a remote storage. After the data is moved, the PVC can be deleted.
```
The files will be stored as sparse qcow2 images.  We will document some examples of how to stitch these files together to construct a restorable image.

aglitke commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+- Handle restart of VM and redefinition of checkpoints.
+- Offline backup
+- API to allow to pull the backup.
+- Handle VM failure where bitmap is not valid - next backup needs to be full.
```
This should be done earlier in my opinion.  Probably during the phase where you handle VM restart and redefinition of the checkpoints since this is the logical place where we will check for dirty bitmap consistency.

ShellyKa13 replied:
sure make sense

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,330 @@
+A `push` mode (where the hypervisor writes out the data to the destination file, which may be local or remote), and a `pull` mode (where the hypervisor creates an NBD server that a third-party client can then read as needed, and which requires the use of temporary storage, typically local, until the backup is complete).
+```
+* For the first phase of implementation, we will start with push mode. In this mode, a PVC will be created and hot-plugged to the virt-launcher. Backup will be initiated, and the data will be pushed to the attached PVC. Afterward, the user can decide what to do with the PVC, and how to move it to a remote location if desired.
+* In the second phase, we will implement pull mode. This will involve defining a complete network API to expose the endpoint that the user can connect to in order to pull the backup. The design for this API is still **TBD** and will be updated accordingly. The endpoint information will be updated in the VMBackup CR for user reference. As specified in libvirt documentation, an additional scratch P ...
````
How so? how will you know what will the user write during the backup?

aglitke replied:
Both.  In the case of push the backup is written to the pvc.  This size is unknown but can be significantly smaller than the total disk size in the case of a small incremental backup.  In the case of pull, there is a scratch disk that maintains a stable view of the data at the time of the backup creation.  This scratch disk will grow each time the VM first changes a block.

pipo replied:
Couple notes:
1) you can query (roughly; it may update) the amount of dirtied blocks for a given backup which gives an estimation on the required size of the (push) backup image

Beware this doesn't apply that for the scratch image, as that holds old versions of blocks which the VM overwrote during pull mode backup. Also note that it happens for all blocks not just those that are covered by the bitmap of dirtied blocks; the NBD export still allows to create a full backup even when an incremental backup is requested.

2) the backup operation will fail if the scratch image runs out of space; a new backup can be taken but obviously it can't be taken at the original time in the past

3) qemu recently introduced a feature that allows discarding blocks which were already read. This could be useful to ensure that the size doesn't bloat if the backup is fast enough. This needs to be explored though.

ShellyKa13 replied:
1. Will you suggest to run a query to get the checkpoint xml with `VIR_DOMAIN_CHECKPOINT_XML_SIZE` flag to get the estimated size before creating the pvc and create one with the size (with a certain buffer)>
2. I understand this is even less accurate in the case of pull, in such case maybe using a bigger buffer should be good enough for most cases and in the case of failure due to the scratch space run out of space to start over with a bigger pvc, as you said it wont be in the exact same time but probably close enough.
3. You mean when using pull mode that the changed blocks during the backup will be discarded for the running backup and will be included in the next one?

pipo replied:
> 1. Will you suggest to run a query to get the checkpoint xml with `VIR_DOMAIN_CHECKPOINT_XML_SIZE` flag to get the estimated size before creating the pvc and create one with the size (with a certain buffer)>
 
Yes you can get a very good estimate from this. In fact for a paused VM (so that no new blocks are dirtied between checking the current size and starting the backup) you could get a 100% accurate prediction for the file size when coulpled with `qemu-img measure` to account for the overhead.

For a running VM you need to include some headroom for the blocks the VM dirties until you manage to start the backup.

>     2. I understand this is even less accurate in the case of pull, in such case maybe using a bigger buffer should be good enough for most cases and in the case of failure due to the scratch space run out of space to start over with a bigger pvc, as you said it wont be in the exact same time but probably close enough.

So for pull mode backups the estimation with `VIR_DOMAIN_CHECKPOINT_XML_SIZE` has exactly the same rules as above. The important fact is that the estimation is for the size of the backup itself.

In pull mode you have also the scratch image which you can't estimate in any way because it depends on how the VM will behave in the future, e.g. how many blocks it will overwrite while the backup job is active. The upper bound is obviously the size of the backed up disk as only the first write to a block by the VM triggers a write (backup of the original state of the block from the main image) to the scratch image.

> 
>     3. You mean when using pull mode that the changed blocks during the backup will be discarded for the running backup and will be included in the next one?

No, what I meant is that this mode (theoretically; I didn't try it) could be used to minimize the size of the scratch image needed for a pull mode backup. The drawback is that each block can be read only once, so the backup application would need to play by the rules otherwise the backup could break if the VM relied on reading it multiple times.

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+* For the first phase of implementation, we will start with push mode. In this mode, a PVC will be created and hot-plugged to the virt-launcher. Backup will be initiated, and the data will be pushed to the attached PVC. Afterward, the user can decide what to do with the PVC, and how to move it to a remote location if desired.
+* In the second phase, we will implement pull mode. This will involve defining a complete network API to expose the endpoint that the user can connect to in order to pull the backup. The design for this API is still **TBD** and will be updated accordingly. The endpoint information will be updated in the VMBackup CR for user reference. As specified in libvirt documentation, an additional scratch P ...
+
+Once the backup is initialized, the controller will pass a backup command to the virt-launcher via the virt-handler using a subresource, containing all the relevant information. Before the backup begins, an FSFreeze command will be issued to ensure file system consistency during the backup. Then, [`virDomainBackupBegin`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainBackupBegin) w ...
```
but if you dont freeze the vm during the backup doesnt it expose a threat to have an inconsistent backup?

aglitke replied:
Yes.  Restoring this type of backup would be like restoring from a power failure but it has the advantage of being a completely transparent backup to the VM.  It's very easy to add an option to quiesce (could be on by default) so I think it's worth having this option.

pipo replied:
Technically even the quiesced filesystem, when restored, qualifies as a recovery from a power outage. It's just that cached writes don't get lost.

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+
+Libvirt has a [Checkpoint](https://libvirt.org/formatcheckpoint.html#checkpoint-xml) resource that represents the point in time when the backup was taken. A checkpoint is created for every backup, including full backups.
+
+During an incremental backup, libvirt returns the changed blocks based on the provided checkpoint name, which is typically the checkpoint from the previous backup. In the `VirtualMachineBackup` CR, the user will add an `incrementalBackup` field, which will be populated with the checkpoint name from which the changed blocks should be tracked.
```
I think this can also be an optional parameter that the backup can provide. Not sure using the exact backup name is good enough since the vmbackup is deleted after the backup one can use the same name for the next backup. we can use the backupname+uid, but I think it will be more valuable to have a defined prefix + the timepstamp of the backup.

pipo replied:
So reallistically the only implication the name has is that is used as the name of the bitmap in qcow2. So you need to be able to reference it later if you want to do a backup.

It needs to be unique only among the active bitmaps in the qcow2 image.

https://www.libvirt.org/kbase/internals/incremental-backup.html#relationships-of-bitmaps-checkpoints-and-vm-disks

Reallistically libvirt will also limit that the checkpoint names are unique per VM, but since the bitmap names are based on the checkpoint name this is effecitvely the same requirement.

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,330 @@
+
+A new field, `QCOW2Overlay`(*NOTE: open to other name suggestions*), will be added to the VirtualMachine and VirtualMachineInstance CRDs. This field will control the addition or removal of the QCOW2 overlay. To enable the feature, one will set it to `true`. If the VM is already running, a restart will be required for the change to take effect (Currently, live addition of the overlay is not possib ...
+ If the VM is already running, a restart will be necessary for the change to take effect (Currently, live addition of the overlay is not possible. Libvirt is working on this, and once it's available, we will work on incorporating it into KubeVirt as well).
+When the VM starts with this field set to `true`, the VMI spec will also be created with this field set to `true`. This will trigger the creation of a VM state PVC (if one doesn’t already exist), create a QCOW2 image overlay on top the raw boot image, and add a data-store tag to the domain XML, resulting in a domain with QCOW2 overlay.
```
Hotplug might be an issue here, since currently you cant do live update of the domain to use the qcow2 overlay so I cant see a reason to create the overlay if it can be used, only after restart this can be done like it is done for any other volumes of the VM..

aglitke replied:
Ahh good point.  Worth mentioning.  So this will be a scenario where a backup would be partially incremental and partially full.  We can cover how the persistently hotplugged disks will be handled though.  In that case, since the disk is present at startup the overlay can be created.

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: my-vm
+    incremental: backup1-16210527022025
```
so the user is responsible for naming checkpoints? Could his be racy? What if the incremental already exists?

Maybe we can come up with a scheme where the user gives a prefix and the checkpoint creation time is appended?

ShellyKa13 replied:
for incremental: "checkpoint-name", this name is the checkpoint from previous backup.
For the checkpointname that should be created for the current backup I wrote:
```
- `checkpointName` <br?>
Optional. If not specified a name will be generated from a prefix of the domain name, followed by the timestamp of the backup. Example: `my-vm-checkpoint-2025-03-03T16:13:28Z`
```

mhenriks replied:
Okay, when I first read this I did not notice that `spec.incremental` and `status. checkpointName` were different.

mhenriks commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,362 @@
+```
+A `push` mode (where the hypervisor writes out the data to the destination file, which may be local or remote), and a `pull` mode (where the hypervisor creates an NBD server that a third-party client can then read as needed, and which requires the use of temporary storage, typically local, until the backup is complete).
+```
+* For the first phase of implementation, we will start with push mode. In this mode, a PVC will either be provided or be created, and will be hot-plugged to the virt-launcher. Backup will be initiated, and the data will be pushed to the attached PVC. Afterward, the PVC will be unplugged and the user can decide what to do with the it, and how to move it to a remote location if desired.
````
One PVC per disk? Or just on PVC?

ShellyKa13 replied:
I thought about 1 PVC for all disks. the PVC will probably need to be Filesystem and get number of files matching the disks backups.

pipo replied:
You definitely want a filesystem. The target image for an *incremental* backup needs to be able to track also the allocation state in addition to the data. This is to be able to properly differentiate a block that was not touched in the time frame the backup is capturing from a block that was e.g. zeroed out.

To use a block device directly you'd most likely need to use QCOW2 in a block device which is really unwieldy. In fact I strongly suggest using qcow2 even as the file format as it tracks allocation state explicitly and is not prone to de-sparsify itself (which equates to the backup being corrupted) with certain operations.

pipo replied:
To clarify. A full backup can be taken into a raw image or block device as all blocks are included ther.

For the incremental backups you need to be able to differentiate blocks which were not changed by the VM since the last backup from the blocks which were zeroed out by the VM. This is represented by allocation state and the unallocated blocks are "transparent" (show the block state from the previous layer, or finally the full backup). This is impossible with block devices. Raw files theoretically can represent allocation state but it can be fragile e.g. if the image desparsifies itself during copy. Thus it's recommended to use qcow2 for push mode backups.

In pull mode, the application pulling the blocks from the NBD server can freely decide what to do with them, e.g. it's out of your scope. The scratch image for pull mode is qcow2 but you should not need to care about that as it's meaningless after the backup.

mhenriks commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,362 @@
+    creationTime: "2025-03-03T16:13:28Z"
+```
+
+Users can list and retrieve checkpoint information via kubectl.
````
Stating the obvious here, but using kubectl will get unwieldy as the number of checkpoints grows. We should think about how users will be querying the checkpoints and optimize for that.

ShellyKa13 replied:
right I mentioned this when I stated the 2 possibilities of storing the checkpoints. Where the second option is not keeping them as CR, store them in the VM state PVC and provide an API that wraps virsh checkpoint commands like list

aglitke replied:
Maybe one CR per VM that has all of the checkpoints in a list?

alicefr replied:
+1 Yes, exactly my comment above

pipo replied:
So reallistically you don't need a previous checkpoint after you successfully back up from it (which creates a new one) and thus can delete it as the backup itself then holds the needed data.

Effectively you thus need 1 checkpoint per backup chain you want to keep. In essence that is usually 1 as people don't tend do backup into multiple places differently.

Only exception would be if you'd have two backup vendor tools doing distinct operations.

ShellyKa13 replied:
@pipo not sure we want to assume the backup provider will not want/need the previous checkpoints and delete them on our initiative.
But I think we got to an agreement that best solution is one resource per VM tracking checkpoints.

pipo replied:
Okay in that case you'll need to be able to keep them around.

mhenriks reviewed (commented):
Still wrapping my head around this, some initial thoughts

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+**Pull mode:**
+Endpoint info is provided in the VirtualMachineBackup CR status. This endpoint will behave as an NBD server and can be read from as long as this is done by a trusted third-party. You should create an NBD client that will read the data, distinguishing between data and zeros, and in the case of incremental backup using the dirty bitmap to map the changed blocks that should be read.
+
```
Not sure I understand what is being served up here. I'm assuming a point in time representation of the disk? How will the user access only the dirty bits? Scanning the entire disk and checking for ones and zeros is what we are trying to avoid

aglitke replied:
It is an nbd server that you can query with qemu-img map and qemu-io.

pipo replied:
The dirty blocks are exposed as an additional bitmap for given export on the NBD server.

The protocol allows querying it.

The blocks exposed are frozen in time at the point the backup was started. The backup tool can fetch any block at their leisure.

ShellyKa13 replied:
I will remove any info related to pull mode and will add it in a subsequent VEP/PR

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+### VM crash
+
+In the event of a VM crash, the CBT information may become corrupted and must be discarded. QEMU provides an API to verify if the dirty bitmaps are valid. Upon VM restart, before redefining the checkpoints, we will check the validity of the bitmaps. If a corrupted bitmap is detected, the checkpoints will be discarded, and the next backup will need to be a full backup.
```
I'll quote qemu doc:
`However, if QEMU does not get a chance to close the file cleanly, the bitmap will be marked as +inconsistent at next load and considered unsafe to use for any operation. At this point, the only valid operation on such bitmaps is block-dirty-bitmap-remove.

Losing a bitmap in this way does not invalidate any existing backups that have been made from this bitmap, but no further backups will be able to be issued for this chain.` 
So it means next backup has to be full and cant use the previous checkpoints to do incremental backups, after the next full backup you will be able to continue to do incremental ones.
Any backups that were performed before the crash can be used to restore the VM since they are not effected.

ShellyKa13 replied:
The backup result is the data itself. The checkpoints are not used for the restore process, they are only used to get the diff in the backup process itself.
Supposedly you can discard any older checkpoint once you create an incremental backup if you want a chain of backups and not a tree.
We need to keep the checkpoints only for the shutdown of the VM so we will be able to do another incremental backup after restart

mhenriks replied:
How does the user know what checkpoint to use? Say they want to restore from the midnght backup? Users are responsible for maintaining a copy of their own metadata?

mhenriks replied:
I guess this info can be stored in the backup CR status?

pipo replied:
(sorry I will not quote everything I'm replying to)

1) A broken bitmap/checkpoint means that qemu forgot which blocks were dirtied since the time the bitmap was created (the time of the last backup). Usually because qemu didn't have time to write out the bitmap.

2) Recovery from this situation is to create another full backup and start a new incremental chain from that point.

3) all previous backups and incrementals (the data itself) is still valid

4) broken bitmaps/checkpoitnts need to be deleted manually. In case you're re-defining it with the flag enabling checking; this will likely mean the need to re-define it without the flag and then delete it once again. Libvirt can consider a mode where it'd delete the bitmaps along the redefinition if they are corrupted. Please file RFE if you want that.

ShellyKa13 replied:
@pipo regarding (4) just clarifying, the steps on restart will be recreate a checkpoint using the xml metadata, checking the matching bitmap is valid. If not delete the checkpoint and all other checkpoints since there will not be useful anymore. There is no way to validate the bitmap before redefining the checkpoint? Cant you query qemu directly the bitmap validity before redefining the checkpoints?

pipo replied:
For libvirt in order to validate the checkpoint it needs to know about as we don't work with bitmaps which do not belong to a checkpoint object.

Libvirt does this by probing qemu so it is possible but for the libvirt API it's a chicken&egg problem. Libvirt also doesn't expose bitmaps itself as they are considered an implementation detail of how backups work.

You also need to have the checkpoint metadata to delete invalid bitmaps due to the same reason as above so I don't see a reason why you'd want to check before defining.

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+            </disks>
+        </domaincheckpoint>
+    OR
+    xmlData: PGRvbWFpbmNoZWNrcG9pbnQ+CiAgICA8bmFtZT4xNTI1ODg5NjMxPC9uYW1lPgogICAgPHBhcmVudD4KICAgICAgICA8bmFtZT4xNTI1MTExODg1PC9uYW1lPgogICAgPC9wYXJlbnQ+CiAgICA8Y3JlYXRpb25UaW1lPjE1MjU4ODk2MzE8L2NyZWF0aW9uVGltZT4KICAgIDxkaXNrcz4KICAgICAgICA8ZGlzayBuYW1lPSd2ZGEnIGNoZWNrcG9pbnQ9J2JpdG1hcCcgYml0bWFwPScxNTI1ODg5NjMxJy8+CiAgICAgICAgPGRpc2sgbmFtZT0ndmRiJyBjaGVja3BvaW50PSdubycvPgogICAgPC9kaXNrcz4KPC9kb2 ...
```
I agree

ShellyKa13 replied:
Do you have an opinion on which way we should keep track of the checkpoints? With a CR for each or storing them in storage, using libvirt api for information?

aglitke replied:
I think a CR will better serve us because we need the API to remove old ones and to represent whether the checkpoint is valid.

mhenriks replied:
If any important info from checkpoints like creation time can be copied to the backup CR and the checkpoints are simply for restarts, I think we may be able to safely stuff them all in a single resource (compressed)

alicefr replied:
I still find an awkward way to put xml into a CR. Don't we have any other options in k8s to store opaque data?

alicefr replied:
Why not an annotation in the VMBackup? Or in the VMBackup status directly

pipo replied:
As noted you effectively only need the name and list of disks included in the checkpoint to reconstruct the XML.

ShellyKa13 replied:
@pipo dont you also need to `parent` of the checkpoint?
@alicefr The VMBackup should be deleted at the end of the backup.
I believe a single resource tracking the backups can be good enough, having the checkpoint names in the spec and the status for them and also maybe the metadata in the status. Then if we want to delete a checkpoint we will watch updates on this resource and compare the map of checkpoints in the spec to the one in the status and update the status accordingly.
What do you think?

pipo replied:
The `parent` field is not strictly required for now. It's used only to be able to query the proper hierarchy but if it's missing the backup will still work.

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+* For the first phase of implementation, we will start with push mode. In this mode, a PVC will either be provided or be created, and will be hot-plugged to the virt-launcher. Backup will be initiated, and the data will be pushed to the attached PVC. Afterward, the PVC will be unplugged and the user can decide what to do with the it, and how to move it to a remote location if desired.
+* In the second phase, we will implement pull mode. This will involve defining a complete network API to expose the endpoint that the user can connect to in order to pull the backup. The design for this API is still **TBD** and will be updated accordingly. The endpoint information will be updated in the VMBackup CR for user reference. As specified in libvirt documentation, an additional scratch P ...
+
+Once the backup is initialized, the controller will pass a backup command to the virt-launcher via the virt-handler using a subresource, containing all the relevant information. Before the backup begins, an FSFreeze command will be issued to ensure file system consistency during the backup. Then, [`virDomainBackupBegin`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainBackupBegin) w ...
```
How about we leave the pull method out of this proposal for now until details are fully fleshed out? Can be added in subsequent VEP?

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+        name: my-vm
+    mode: push
+status:
+    checkpointName: backup1-16210527022025
```
Wondering if it could make sense to keep the checkpoint resources as an implementation detail and have users use a previous backup name here or empty for full backup?

ShellyKa13 replied:
The idea was the once backup is done and the user completed fetching it, it would delete the backup CR since there is no use to it afterwards.. so using a previous backup name doesnt really matter which one is it.. and also as mentioned the backup name doesnt have to be unique since it is being deleted once backup is over..
Regarding empty or provide a name is adding the incremental or not, that decides to use the given name as the checkpoint to backup from or if incremental wasnt provided it will be full.
We dont really have to put the checkpoint name in the status but if the user doesnt want to pick the checkpoint name for current backup it can just be generated and thats the way to update the generated value

mhenriks replied:
I don't think users should have to be concerned with resources that they did not create. For example, users of the `VolumeSnapshot` API can be totally oblivious of `VolumeSnapshotContent` even though the `VolumeSnapshotContent` is technically the resource that contains all the metadata to create a PVC from a snapshot. I think the Backup/Checkpoint relationship should be similar. 

Think of it from a user perspective. They created a backup in the morning called `wednesday_morning`. Now it's the afternoon and they want to do an incremental. With the proposed scheme, they have to get the checkpoint name from the `wednesday_morning` backup. Wouldn't it be easier to just use the previous backup name that you have in your head?

Users are still free to delete backups that they are confident will never be referred to again. But the idea of creating a backup and instantly deleting it doesn't make a lot of sense to me. To do another incremental, the user would have to list all the checkpoints to find the most recent.

mhenriks replied:
I like the one to one relationship between Backup and Checkpoint better than the alternatives. I simply think users should not have to know/be aware of the fact that the Checkpoint resource exists.

@aglitke @alicefr what do you think?

alicefr replied:
All the relationship are still a bit confusing to me, but I personally would try to keep the API as simple as possible, not necessary stick to libvirt APIs. Libvirt APIs are often very complex, with a lot of option to tune, and rather complex to understand.

alicefr commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+This overlays will be stored on a VM state PVC, similar to how TPM and EFI are handled. If the VM uses either of these features, the overlays will be created on the same PVC. Figure 1 describes how this will look like:
+![figure 1](qcow2overlay.png)
+
+A new field, `CBT`(*NOTE: open to other name suggestions*), will be added to the VirtualMachine and VirtualMachineInstance CRDs. This field will control the addition or removal of the QCOW2 overlay to each of the VM disks. To enable the feature, one will set it to `true`. If the VM is already running, a restart will be required for the change to take effect (Currently, live addition of the overla ...
```
What about `backup`?

ShellyKa13 replied:
I think `backup` might be confusing and might cause someone to think it will trigger backup. Also when you set this the actual action it will do is add the qcow2 in order to enable CBT not backup per-se

alicefr commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+### Offline backups
+
+Another challenge that needs to be addressed is taking offline backups. Currently, there is no option to create offline backups in Libvirt and QEMU. Backups require QEMU, as it is necessary to interpret the QCOW2 metadata/bitmap.
```
Any chance to use libguestfs instead? I look like a good fit for tekton tasks or k8s job. Can we maybe put this into a separate proposal. This looks already quite dense

pipo replied:
No. 'libguestfs' is used to access *files* from *inside* the VM. The backup is dealing with *blocks* from *outside* of the VM.

To interpret incremental backups you need a qcow2-aware tool. Only applicable options are `qemu`, `qemu-nbd`, and `qemu-storage-daemon`.

Another limiting factor is that a qcow2 image can't be opened from 2 processes if one of them wants to write into the image. Thus in case when `qemu-nbd` or `qemu-storage-daemon` would be used you'd be unable to start the VM until that process terminates.

alicefr replied:
We discussed this offline, but let me express here again my concerns. 

In general, I'm not against pausing the guest, however, this is as some consequences in how KubeVirt needs to create the launcher pod.
Usually, when the VM is offline, the resources required by the VM aren't allocated (cpu, memory, devices, network...). Now, if we need to spin up the launcher pod for performing the offline backup, then we need to allocate all the other resources, that might not be required for just the backup job. 

It will be much better to have a lighter way to allocate the resources only needed by the back up. However, how can we understand that Kubevirt needs to define another pod with less resource? On top of this, today, we generate the libvirt xml based on the VM definition. If we need to create the a paused guest, then we should probably define the xml with only the required volumes to backup.

pipo replied:
So I think this needs to be approached from the view of how Kubervirt wants to behave in cases when the lifecycle of the VM does intersect with the lifecycle of the backup job.

Just then we can discuss approaches and optimizations.

While the backup jobs is within the boundaries of the VM lifecycle everything is well and we could optimize many things.

If the backup job intersects the boundaries of the VM lifecycle though some concesions need to be made. You'll either need to inconvenience the user, inconvenience the backup or throw more resources at it.

Thus in situations like (below) you can:

1) VM shuts down while backup is running
 
1.1) Inconvenience the backup - stop the backup job
1.2) Throw resources at it - keep the VM around in stopped state, allowing the backup to finish

2) VM wants to start while (offline) backup is running

2.1) Inconvenience the user - wait until the backup finishes; start afterwards
2.2) Inconvenience the backup - kill the backup; start the VM 
2.3) Throw more resources at it - start the offline backup job as a paused VM

Now once you clarify what the desired concession in the above cases is we can talk about implementation and possible limitations.

ShellyKa13 replied:
@pipo I think 2.3 should be a different section completely since it doesn't necessarily mean that the VM will be started during the backup. By offline backups I mean no changes to the VM state are required - we dont really want to start the VM.
Since in our offline discussions it doesnt seem like there is a convenient way to do an offline backup that will be consistent with the online ones.
@alicefr I agree that this VEP is already pretty dense and I agree that offline backup can be done in a subsequent PR. I will mention this in the VEP and we can have a discussion on this in the future. Having an online backup seems like a valid limitation as a first phase.

pipo replied:
> I think 2.3 should be a different section completely since it doesn't necessarily mean that the VM will be started during the backup.

My intention of the whole above reply was to clarify what Kubevirt wants to do *if* the backup job lifetime doesn't align with the VM lifetime.

This was not yet clarified.

Point 2.3 is a legitimate approach *if* you want to allow users to start the VM. It's also an approach if you want to use the same interface with libvirt.

> By offline backups I mean no changes to the VM state are required - we dont really want to start the VM.

Okay so that is technically point 2.1 (at least when going offline->online); disallow the user change the state. 

> Since in our offline discussions it doesnt seem like there is a convenient way to do an offline backup that will be consistent with the online ones.

Based on few of the other comments you seem to be focusing on *push* mode backups. Note that the "simple" tools (`qemu-nbd`, `nbdcopy`, `nbdkit`, `qemu-img`) don't actually provide a way to do a "push" backup as qemu does. It will require either to run `qemu` itself or `qemu-storage-daemon`.

Anyways; the above is an implementation detail. First and foremost you should decide what to do *if* the user will want to start the VM while the backup is running. We can then discuss how to do it the simplest way.

alicefr commented:
@ShellyKa13 I have a couple of questions:
1. Why cannot we simply store everything in `VirtualMachineBackup`? 
2. The Checkpoint CR will require somebody to creates/updates it? VIrt-launcher cannot access kubernetes api, so, do you plan to add a new GRPC call between the handler and the launcher? Or how this should work?
3. Do we really need to extend the VM/VMI spec? Cannot we simply list the volumes to backup in the `VirtualMachineBackup`? Otherwise, how does that field work with live-updates? If I set a volume with `CBT`: `true`, does it mean it is immutably part of the backup?

pipo commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+Once the backup is initialized, the controller will pass a backup command to the virt-launcher via the virt-handler using a subresource, containing all the relevant information. Before the backup begins, an FSFreeze command will be issued to ensure file system consistency during the backup. Then, [`virDomainBackupBegin`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainBackupBegin) w ...
+
+Once the data has been pushed or fetched, the VirtualMachineBackup resource can be deleted. Deleting the object will abort the libvirt backup job, effectively terminating the backup process.
```
Note that for push-mode the backup job in libvirt terminates itself as soon as everything is backed up. The client application is notified via an event about this.

Aborting/terminating the job is needed only in pull-mode.

ShellyKa13 replied:
Yes, I will add a clarification for that. Thanks

pipo commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+Libvirt has a [Checkpoint](https://libvirt.org/formatcheckpoint.html#checkpoint-xml) resource that represents the point in time when the backup was taken. A checkpoint is created for every backup, including full backups. The checkpoint name can be provided, or if not, it will automatically use a predefined prefix followed by the backup timestamp.
+
+During an incremental backup, libvirt returns the changed blocks based on the provided checkpoint name, which is typically the checkpoint from the previous backup. In the `VirtualMachineBackup` CR, the user will add an `incrementalBackup` field, which will be populated with the checkpoint name from which the changed blocks should be tracked.
+In push mode, the data pushed to the PVC will include only the data changes up to the backup point. In pull mode, the user will be asked to request only the dirty bitmap blocks.
```
As noted above in pull mode the application can still theoretically do a full backup even when incremental mode is enabled. At least with the qemu NBD connection.

The blocks to be backed up are exposed as a bitmap associated with the NBD export, which also exposes the full state of the block device at the time the backup was started. The client then drives what it pulls.

pipo commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+Two approaches are being considered to be able to redefine the checkpoints after a VM restart:
+- Create a `Checkpoint` CR to store the XML data, either as a plain string or base64 encoded. This approach is more visible and easier to query using the k8s API, but could result in a large number of checkpoints per VM, potentially causing performance issues due to the increased number of objects in the cluster.
+- Store checkpoint metadata alongside the QCOW2 overlay in the VM state PVC to ensure persistence. The checkpoint will be stored as an XML, which takes up very little space. This method does not introduce another CR, but since it is not a Kubernetes resource, it requires an alternative method to get checkpoints information. To access the existing checkpoints data, it will be possible to either ex ...
```
Effectively; you need to store only the name of the checkpoint and the list of disks it was applied to. It's relatively easy to re-create the XML from the information above.

You can store additional metadata yourself if you don't want to use the libvirt APIs for those.

Dumping the libvirt XML; storing it and using it obviously works too, but is not strictly needed.

pipo commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+According to the previous solution, if the VM was offline when the backup was requested, we will start the VM in a paused state. When the VM is in a paused state, there is no issue with starting it during the backup process, as Libvirt and QEMU guarantee that any VM operations, such as overwriting blocks in the data file, will not affect the backup's consistency. This consistency assurance could  ...
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup or before stopping the VM we would wait for the VirtualMachineBackup object to be deleted and only the ...
```
Beware that the guest OS can initiate shutdown itself too which you really can't prevent. 

The libvirt solution will in such case also transition the VM to paused state. My plan is to also have a special flag for the backup job so that the paused state will linger only while the backup persists.

ShellyKa13 replied:
@pipo Sounds good, obviously a build it solution by you guys will be the best :)
In case of guest OS shutdown, as you said not really sure what can be done. In such case will the backup job simply fail?

pipo replied:
The job will be killed along with the VM. This means that for a push mode the image will be incomplete (and possibly even deleted). For pull mode the NBD server will terminate.

One of the RFE's I've filed myself for libvirt is to add a flag for the backup job to allow the VM linger until the job finishes which would fix this lifecycle tansition.

pipo commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+
+Once the raw restored image is created, you can store it in a PVC and use this PVC as the restored volume for the VM. You can use any population method like import or upload.
+
+
```
The above is a good example with qcow2 images, but I expect the backup vendor to have some internal magic in it's tool that effectively spits out the image afterwards.

pipo replied:
Actually I've noticed that in the example the images are not layered up properly:

```bash
$ qemu-img rebase -b fullbackup.qcow2 -f qcow2 -u incremental2.qcow2
$ qemu-img rebase -b fullbackup.qcow2 -f qcow2 -u incremental3.qcow2
.

$ qemu-img convert -f qcow2 -O raw fullbackup.qcow2 restored-raw.img
```

The step layering up `incremental2.qcow2` on top of `fullbackup.qcow2` is correct.

You then need to layer up `incremental3.qcow2` on top of `incremental2.qcow2`.

Also the coversion to a full flattened raw image needs to then be done from `incremental3.qcow2`

pipo commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+Disadvantages:
+- This KEP is still in the early stages of implementation and will take considerable time to mature before reaching a stable v1 version and it is not under our control.
+- Beyond waiting for the Kubernetes API to mature, each storage provider must implement the optional SnapshotMetadata API as outlined in the KEP. Since this is optional, not all providers will support it, and adoption will be gradual, restricting the available options for users.
+- When using Kubernetes VolumeSnapshot, the guest must be frozen during the snapshot of all the VM volumes. In contrast, libvirt uses a dedicated job to capture the disk state quickly, minimizing guest downtime.
```
This seems to be an advantage.

ShellyKa13 replied:
this is an advantage for the current VEP with libvirt, disadvantage for the k8s KEP

pipo reviewed (commented):
Sorry for getting in late, I learned about this thread only yesterday.

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+A challenge in Kubevirt is that Libvirt is recreated each time a VM restarts, causing any existing checkpoints to be lost. To continue incremental backups after a restart, the checkpoint metadata must be provided to libvirt during VM initialization. This allows libvirt to redefine the checkpoints, ensuring the incremental backup process can resume from the correct point in time.
+
+Two approaches are being considered to be able to redefine the checkpoints after a VM restart:
+- Create a `Checkpoint` CR to store the XML data, either as a plain string or base64 encoded. This approach is more visible and easier to query using the k8s API, but could result in a large number of checkpoints per VM, potentially causing performance issues due to the increased number of objects in the cluster.
```
Yes thats another possibility that can be considered, I'll add it with its pros and cons also..

ShellyKa13 replied:
I removed the last option since I think its the least useful and most complex. Instead added this option which I think might scale the best and is not as complex as the other option.

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,362 @@
+    </disk>
+```
+
+In the case the field is not present or set to `false`, the VM state PVC will deleted if it is not required by other features. The QCOW2 images will be deleted if exists. And the domain XML will be modified back to use the raw images as the disks images.
````
yes, can be turned on with true and off with false/deletion

ShellyKa13 commented:
@aglitke @mhenriks @alicefr @pipo Updated the VEP.
Removed information on pull mode and mentioned it will be in a subsequent PR/VEP.
Updated the tracking of checkpoints to be with a single resource.
Will appreciate another look.

mhenriks commented:
@ShellyKa13 it appears that the VEP is unchanged from previous commit
https://github.com/kubevirt/enhancements/compare/a14712a67fe141c37e7661b9620ad68f7c6ae87a..07a0518453babba13e6126c935f8dc80ed408922

ShellyKa13 commented:
@mhenriks oops, thanks, not should include the changes

alicefr commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,362 @@
+This overlays will be stored on a VM state PVC, similar to how TPM and EFI are handled. If the VM uses either of these features, the overlays will be created on the same PVC. Figure 1 describes how this will look like:
+![figure 1](qcow2overlay.png)
+
+A new field, `changedBlockTracking`(*NOTE: open to other name suggestions*), will be added to the VirtualMachine and VirtualMachineInstance CRDs. This field will control the addition or removal of the QCOW2 overlay to each of the VM disks. To enable the feature, one will set it to `true`. If the VM is already running, a restart will be required for the change to take effect (Currently, live addit ...
```
Can we create the overlay by default or are there any performance penalty?

alicefr replied:
Would this avoid any needs of an API addition

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,362 @@
+    pvcName: my-vm-backup-2025-03-03T16:13:28Z
+```
+
+### VirtualMachineCheckpointsList CR
````
No. There are valid checkpoints since the first full backup. Most users will probably not use those since they will probably always do an incremental backup from the latest one but it is possible to do one also from older once.
Also as @mhenriks commented its possible there will be several backup solutions running for the same VM so one needs to be able to use the proper checkpoint for the matching backup chain.

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+        name: my-vm
+    checkpointsNames:
+    - my-backup-chain-2025-03-03T16:13:28Z
+    - my-backup-chain-2025-03-03T17:13:28Z
```
`checkpointNames` is a better name but why are these here in the first place? The spec should only be for things that the user can specify. Isn't `status.checkpoints` sufficient?

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+This field will be set to `true` either:
+* Automatically, when the first backup is initiated.
+* Automatically, upon creation of `VirtualMachineBackupChain` CR (will be explained in next section)
+* Manually, by the user — helps avoid a restart right before the first backup and reduces the time required to initiate it.
```
Maybe only support manual initially? I assume you are already planning to have conditions on backup and backup chain for the case when they are created before this flag is set

ShellyKa13 replied:
That was the original direction but @alicefr prefers not having it at all, so I thought the middle ground was to have it either be automatically so the user doesnt really have to know this added API, or manually if the user do want to explicitly set it and have it ready for backups upfront.

alicefr replied:
This isn't completely, my point was to  avoid any API additions on the VM spec. I don't thin we can support anything which automatically set the field in behalf of the users. If the VM is handled with gitops, then if there is a controller monitoring the VM changes, it will detect the spec change, compare it with the original VM definition and revert your change.

alromeros replied:
In that case the best option might be to always ask users to set it manually in advance. Not having this behavior clearly defined in the VM spec might be problematic when checking if a VM is using a persistent state PVC, which is necessary during flows like VM snaphsot and restore or backup. There's no direct API to specify the creation of a backend storage PVC, but we can usually infer if one's been created by checking if persistent TPM or EFI are set to true. From now on checking `changedBlockTracking` will be also required to see if this PVC has been created. If the field can be false (or nonexistent) in the spec but still set internally to true by Kubevirt according to some other logic, the VM might be using a backend storage PVC without any comprehensible way for us to know, which has been proven to be problematic. Just something to consider.

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+3. A `data-store` tag will be added to each of those disks in the domain XML.
+As a result, the domain will be configured with QCOW2 overlays.
+Once that is done the VMI volume status will be updated with changedBlockTracking: `true` for the relevant volumes.
+*NOTE: In the case of a hot-plugged disk, the VM will need to be restarted for it to be included in the backup and have CBT. Once live addition of the overlay is supported, restarts will no longer be required.
```
What happens when backup is created with hotplug volume and no overlay? I assume the backup will proceed but there will be some indication in the backup status?

ShellyKa13 replied:
Hmm, well in the volume status of the vmi you can see which disks has the overlay.
Do we want to have `includedVolumes` list similar to in vmsnapshot that will indicate which volumes were included? 
Also in the checkpoints list info if it will not be encoded you can see that xml of the checkpoint which states which disks were included in the backup, but not sure if it will not make the backupChain status too big

ShellyKa13 replied:
As discussed offline, we dont have the problem of a restart with hotplug since we can create the overlay dynamically when hotplugging the volume, so no need for this field or indication

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+
+### State interruptions during backup
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup or before stopping the VM we would wait for the VirtualMachineBackup job to complete and only then all ...
```
Is this fix going to be ready before this feature ships? My concern is that, unlike with VM snapshots, a backup may take a while, especially a full one. Is there a way to gracefully cancel the backup and leave the bitmap in a good state for when the VM comes back up?

ShellyKa13 replied:
@pipo correct me if Im wrong, there is an option to send an abort action to the backup job, and we can do it in case the backup object is deleted before the backup is `done`.
Regarding the bitmaps Im not sure what will be the state of it, if the backup never completed will it be deleted automatically, or will need to also delete it after aborting the backup, and then shutdown the VM should be good.

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+Migration and backup operations are mutually exclusive and cannot be performed simultaneously.
+During migration, the bitmaps of all previous backups are automatically transferred along with the disk image. The only additional step required is similar to what is done on VM restarts: the checkpoints must be redefined for the newly initialized Libvirt instance.
+TODO: Confirm whether the VM state PVC also needs to be migrated.
+If migration of the state PVC is needed but not possible, all existing checkpoints will be lost, and the next backup will have to fall back to a full backup.
```
Can we be more clear about what happens when a VM is migrated during backup? I assume the backup is cancelled/aborted? Even if migration can theoretically work while backing up, user may not be backing up to RWX PVC. Given that Filesystem PVC is required for backup I expect that to be the most common case

ShellyKa13 replied:
I assume that migration has a bigger priority? if so I guess backup will be aborted.

> Even if migration can theoretically work while backing up, user may not be backing up to RWX PVC. Given that Filesystem PVC is required for backup I expect that to be the most common case
Not sure I understand the question here. From what I understad from @pipo we dont really care if the storage and VM state pvc is RWX since we can just create a new one in the migration target, the data will be kept in the raw image and the bitmaps will be streamed in the qemu migration

mhenriks commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,374 @@
+    mode: push
+status:
+    checkpointName: my-vm-backup-2025-03-03T16:13:28Z
+```
````
add `pvcName`?

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+The most recent checkpoint created, updated by the backup controller. This checkpoint will serve as the base for the next incremental backup.
+- `checkpoints`<br>
+A list of checkpoint entries. Each entry includes the checkpoint's name, creation timestamp, and metadata. Metadata is base64-encoded to prevent YAML clutter.
+
```
What about conditions? Off the top of my head, should indicate when VM must be rebooted, also some indication when all the bitmaps are corrupted and everything is cleared out

ShellyKa13 replied:
I'm not sure the indication that the VM needs to be rebooted fits in the backupChain conditions. There is a condition like that in the VM "restart required" for any changed in the spec that are not live updates - which I think should be good here.
Regarding the corrupted bitmaps, when those are met in restart I think the checkpoints will just get deleted so not sure there will be enough time to indicate that, once they are deleted the chain should just be empty.

mhenriks commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+- `checkpointName`<br>
+The name of the checkpoint created for the current backup. This field is updated by the backup controller.
+- `conditions`<br>
+Represents the current state of the backup process, such as `Initializing`, `Done`, `Failed` or `Deleting`
```
Condition if some hotplug PVCs are excluded?

mhenriks reviewed (commented):
Getting there! I like the new API

Currently mostly wondering about interaction with live migration and how to handle VM shutdown

alicefr commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,374 @@
+
+A new field, `changedBlockTracking`, will be added to the VirtualMachine CRD. This field will control the addition or removal of the QCOW2 overlay for each of the supported VM volumes (i.e DataVolumes and PersistentVolumeClaims).
+This field will be set to `true` either:
+* Automatically, when the first backup is initiated.
```
This isn't git ops compatible, we cannot change the VM spec under the hood

mhenriks commented on veps/sig-storage/incremental-backup.md:
````diff
@@ -0,0 +1,374 @@
+
+*VirtualMachineBackupChain
+```yaml
+apiVersion: kubevirt.io/v1
````
different api group

alicefr commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,420 @@
+    latestCheckpoint:
+      name: my-backup-tracker-2025-03-03T16:13:28Z
+      creationTime: "2025-03-03T16:13:28Z"
+      xmlData: |
```
Why do we need the xmlData in the status?

ShellyKa13 replied:
We need the checkpoint metadata to be able to redefine them, but I like your direction of just wrap libvirt checkpoint xml as a go struct and have it as the latestCheckpoint explicitly. Ill update the VEP accordingly.

alicefr commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,420 @@
+    latestCheckpoint:
+      name: my-backup-tracker-2025-03-03T16:13:28Z
+      creationTime: "2025-03-03T16:13:28Z"
+      xmlData: |
```
If this is really necessary why don't we simply use proper structs instead of some unstructured data

alicefr commented:
@ShellyKa13 still a couple of questiosn:
1. What happen if we have a VirtualMachineBackup which reference an invalid VirtualMachineBackupTracker? 
2. Are VirtualMachineBackups immutable? I would guess so

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,420 @@
+
+To reflect the changedBlockTracking state, an indication will be added to the status of selected VirtualMachines. This state can be:
+- `PendingRestart`: The VM needs a restart for the changes to take effect.(will not be needed once libvirt enhancement is available)
+- `Initializing`: A restart has occurred, and the necessary setup for the changes is in progress.
```
I figured we want to indicate the user that the restart was made, but until the overlay is set and vmi is ready then the feature is not technically enabled. Also currently the backup will not be supported until the vmi is running so it seems the state is compatible with the vmi status after restart(until running backup is not available anyways) but when offline will be available, until the overlay exists you cant actually do the backup so youll want some indication for that too..

ShellyKa13 commented on veps/sig-storage/incremental-backup.md:
```diff
@@ -0,0 +1,420 @@
+
+NOTE: For hot-plugged disks, a restart is not necessary as they can be hot-plugged with the data-store tag already present in the domain XML.
+
+If a VM no longer matches the label selector (after a restart), all related resources will be removed, the `changedBlockTracking` field will be removed from the VMI volume status, and the `changedBlockTracking` field in the VM status will be set to `disabled`.
```
yeah I thought to use the pendingRestart also in the case of disabling the CBT. I can add it to the VEP

alicefr commented:
Looks good to me, great work @ShellyKa13!
/approve

mhenriks commented:
/lgtm

ShellyKa13 commented:
I rebased and removed a duplicate line: [veps/sig-storage/incremental-backup.md](https://github.com/kubevirt/enhancements/compare/bbf542a49ac0dd80d536cf3ef7cc3d9e0c6d553d..1d71eb14084d2edeb95ec4c99be33e49135e7a3b#diff-d71144101f7b40d3ea0679b0815598ab2a5302f80aa44d54060c3d45867c3121)
lgtm was removed but the VEP is the same :)

ShellyKa13 commented:
/hold
I want to make another small change

vladikr commented:
Thank you @ShellyKa13 , looks good from my pov.
/approve

mhenriks commented:
/lgtm

alicefr commented:
/lgtm
/unhold
